File: Neato-SDK/neato-sdk-android/src/main/java/com/neatorobotics/sdk/android/robotservices/cleaning/CleaningService.java
package com.neatorobotics.sdk.android.robotservices.cleaning;

import android.util.Log;

import com.neatorobotics.sdk.android.NeatoCallback;
import com.neatorobotics.sdk.android.NeatoError;
import com.neatorobotics.sdk.android.models.Robot;
import com.neatorobotics.sdk.android.models.RobotState;
import com.neatorobotics.sdk.android.nucleo.NucleoResponse;
import com.neatorobotics.sdk.android.nucleo.RobotConstants;
import com.neatorobotics.sdk.android.robotservices.RobotService;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

/**
 * Neato
 * Created by Marco on 05/01/2017.
 * Copyright © 2016 Neato Robotics. All rights reserved.
 */

public abstract class CleaningService extends RobotService {

    private static final String TAG = "CleaningService";

    protected int cleaningType = RobotConstants.ROBOT_CLEANING_CATEGORY_HOUSE;

    public abstract boolean isEcoModeSupported();
    public abstract boolean isTurboModeSupported();
    public abstract boolean isExtraCareModeSupported();
    public abstract boolean isCleaningAreaSupported();
    public abstract boolean isCleaningFrequencySupported();
    public abstract boolean isFloorPlanSupported();
    public abstract boolean areZonesSupported();

    public void stopCleaning(final Robot robot, HashMap<String,String> params, final NeatoCallback<RobotState> callback) {
        JSONObject command = new JSONObject();
        try {
            command.put("reqId", "77");
            command.put("cmd", "stopCleaning");
        }catch (JSONException e) {
            Log.e(TAG, "Exception", e);
        }

        nucleoClient.sendCommandAsync(robot, command, new NeatoCallback<NucleoResponse>(){
            @Override
            public void done(NucleoResponse result) {
                super.done(result);
                RobotState robotState = new RobotState(result.getJSON());
                callback.done(robotState);
            }

            @Override
            public void fail(NeatoError error) {
                super.fail(error);
                callback.fail(error);
            }
        });
    }

    public void pauseCleaning(final Robot robot, HashMap<String,String> params, final NeatoCallback<RobotState> callback) {
        JSONObject command = new JSONObject();
        try {
            command.put("reqId", "77");
            command.put("cmd", "pauseCleaning");
        }catch (JSONException e) {
            Log.e(TAG, "Exception", e);
        }

        nucleoClient.sendCommandAsync(robot, command, new NeatoCallback<NucleoResponse>(){
            @Override
            public void done(NucleoResponse result) {
                super.done(result);
                RobotState robotState = new RobotState(result.getJSON());
                callback.done(robotState);
            }

            @Override
            public void fail(NeatoError errorCode) {
                super.fail(errorCode);
                callback.fail(errorCode);
            }
        });
    }

    public void resumeCleaning(final Robot robot, HashMap<String,String> params, final NeatoCallback<RobotState> callback) {
        JSONObject command = new JSONObject();
        try {
            command.put("reqId", "77");
            command.put("cmd", "resumeCleaning");
        }catch (JSONException e) {
            Log.e(TAG, "Exception", e);
        }

        nucleoClient.sendCommandAsync(robot, command, new NeatoCallback<NucleoResponse>(){
            @Override
            public void done(NucleoResponse result) {
                super.done(result);
                RobotState robotState = new RobotState(result.getJSON());
                callback.done(robotState);
            }

            @Override
            public void fail(NeatoError errorCode) {
                super.fail(errorCode);
                callback.fail(errorCode);
            }
        });
    }

    public void returnToBase(final Robot robot, HashMap<String,String> params, final NeatoCallback<RobotState> callback) {
        JSONObject command = new JSONObject();
        try {
            command.put("reqId", "77");
            command.put("cmd", "sendToBase");
        }catch (JSONException e) {
            Log.e(TAG, "Exception", e);
        }

        nucleoClient.sendCommandAsync(robot, command, new NeatoCallback<NucleoResponse>(){
            @Override
            public void done(NucleoResponse result) {
                super.done(result);
                RobotState robotState = new RobotState(result.getJSON());
                callback.done(robotState);
            }

            @Override
            public void fail(NeatoError errorCode) {
                super.fail(errorCode);
                callback.fail(errorCode);
            }
        });
    }

    public void startCleaning(final Robot robot, HashMap<String, String> params, final NeatoCallback<RobotState> callback) {
        // default values
        int cleaningMode = RobotConstants.ROBOT_CLEANING_MODE_TURBO;

        int cleaningModifier = RobotConstants.ROBOT_CLEANING_MODIFIER_NORMAL;
        int cleaningNavigationMode = RobotConstants.ROBOT_EXTRA_CARE_MODE_OFF;
        int cleaningAreaWidth = RobotConstants.ROBOT_CLEANING_SPOT_SIZE_LARGE;
        int cleaningAreaHeight = RobotConstants.ROBOT_CLEANING_SPOT_SIZE_LARGE;

        // override default values, with user values
        // discard unsupported values
        if(params != null) {
            if(params.containsKey(RobotConstants.CLEANING_MODE_KEY)) {
                cleaningMode = Integer.parseInt(params.get(RobotConstants.CLEANING_MODE_KEY));
            }
            if(params.containsKey(RobotConstants.CLEANING_MODIFIER_KEY)) {
                cleaningModifier = Integer.parseInt(params.get(RobotConstants.CLEANING_MODIFIER_KEY));
            }
            if(params.containsKey(RobotConstants.CLEANING_NAVIGATION_MODE_KEY)) {
                cleaningNavigationMode = Integer.parseInt(params.get(RobotConstants.CLEANING_NAVIGATION_MODE_KEY));
            }
            if(params.containsKey(RobotConstants.CLEANING_AREA_SPOT_HEIGHT_KEY)) {
                cleaningAreaHeight = Integer.parseInt(params.get(RobotConstants.CLEANING_AREA_SPOT_HEIGHT_KEY));
            }
            if(params.containsKey(RobotConstants.CLEANING_AREA_SPOT_WIDTH_KEY)) {
                cleaningAreaWidth = Integer.parseInt(params.get(RobotConstants.CLEANING_AREA_SPOT_WIDTH_KEY));
            }
        }

        JSONObject command = new JSONObject();
        try {
            command.put("reqId", "77");
            command.put("cmd", "startCleaning");

            JSONObject commandParams = new JSONObject();
            commandParams.put(RobotConstants.CLEANING_CATEGORY_KEY, cleaningType);
            if(isEcoModeSupported() && isTurboModeSupported()) {
                commandParams.put(RobotConstants.CLEANING_MODE_KEY, cleaningMode);
            }
            if(isCleaningFrequencySupported()) {
                commandParams.put(RobotConstants.CLEANING_MODIFIER_KEY, cleaningModifier);
            }
            if(isExtraCareModeSupported()) {
                commandParams.put(RobotConstants.CLEANING_NAVIGATION_MODE_KEY, cleaningNavigationMode);
            }
            if(isCleaningAreaSupported()) {
                commandParams.put(RobotConstants.CLEANING_AREA_SPOT_HEIGHT_KEY, cleaningAreaHeight);
                commandParams.put(RobotConstants.CLEANING_AREA_SPOT_WIDTH_KEY, cleaningAreaWidth);
            }
            // check if floor plan cleaning is supported
            // and if we have some floor plan or zones related params
            if(isFloorPlanSupported()) {
                if(params.containsKey(RobotConstants.CLEANING_TYPE_KEY)) {
                    cleaningType = Integer.parseInt(params.get(RobotConstants.CLEANING_TYPE_KEY));
                    commandParams.put(RobotConstants.CLEANING_CATEGORY_KEY, cleaningType);// override house with floor plan if the case
                }
            }
            if(areZonesSupported()) {
                if(params.containsKey(RobotConstants.CLEANING_ZONE_KEY)) {
                    String boundaryId = params.get(RobotConstants.CLEANING_ZONE_KEY);
                    commandParams.put(RobotConstants.CLEANING_ZONE_KEY, boundaryId);
                }
            }
            command.put("params", commandParams);
        }catch (JSONException e) {
            Log.e(TAG, "Exception", e);
        }

        nucleoClient.sendCommandAsync(robot, command, new NeatoCallback<NucleoResponse>(){
            @Override
            public void done(NucleoResponse result) {
                super.done(result);
                RobotState robotState = new RobotState(result.getJSON());
                callback.done(robotState);
            }

            @Override
            public void fail(NeatoError errorCode) {
                super.fail(errorCode);
                callback.fail(errorCode);
            }
        });
    }
}

File: Neato-SDK/neato-sdk-android/src/main/java/com/neatorobotics/sdk/android/robotservices/cleaning/CleaningService.kt
/*
 * Copyright (c) 2019.
 * Neato Robotics Inc.
 */

package com.neatorobotics.sdk.android.robotservices.cleaning

import android.util.Log

import com.neatorobotics.sdk.android.clients.nucleo.Nucleo
import com.neatorobotics.sdk.android.robotservices.RobotService
import com.neatorobotics.sdk.android.clients.Resource
import com.neatorobotics.sdk.android.clients.nucleo.NucleoRepository
import com.neatorobotics.sdk.android.models.*
import com.neatorobotics.sdk.android.robotservices.housecleaning.HouseCleaningBasic1Service
import org.json.JSONArray

import org.json.JSONException
import org.json.JSONObject

import java.util.HashMap

val Robot.cleaningService: CleaningService
    get() {return HouseCleaningBasic1Service()}

abstract class CleaningService : RobotService() {

    // default values
    protected var cleaningType = CleaningCategory.HOUSE

    protected var cleaningMode = CleaningMode.TURBO
    protected var cleaningModifier = CleaningModifier.NORMAL
    protected var cleaningNavigationMode = NavigationMode.NORMAL
    protected var cleaningAreaWidth = Nucleo.SPOT_AREA_LARGE
    protected var cleaningAreaHeight = Nucleo.SPOT_AREA_LARGE
    protected var boundaryId = ""
    protected var boundaryIds = listOf<String>()

    abstract val isEcoModeSupported: Boolean
    abstract val isTurboModeSupported: Boolean
    abstract val isExtraCareModeSupported: Boolean
    abstract val isCleaningAreaSupported: Boolean
    abstract val isCleaningFrequencySupported: Boolean
    abstract val isMultipleZonesCleaningSupported: Boolean

    suspend fun stopCleaning(robot: Robot, params: HashMap<String, String>? = null): Resource<RobotState> {
        val command = JSONObject().apply {
            put("reqId", "77")
            put("cmd", "stopCleaning")
        }

        val result = nucleoRepository.executeCommand(robot, command)

        val newState = RobotState(serial = robot.serial).apply {
            loadData(result.data)
        }

        return when(result.status) {
            Resource.Status.SUCCESS -> Resource.success(newState)
            else -> Resource.error(result.code, result.message, newState)
        }
    }

    suspend fun pauseCleaning(robot: Robot, params: HashMap<String, String>? = null): Resource<RobotState> {
        val command = JSONObject().apply {
            put("reqId", "77")
            put("cmd", "pauseCleaning")
        }
        val result = nucleoRepository.executeCommand(robot, command)

        val newState = RobotState(serial = robot.serial).apply {
            loadData(result.data)
        }

        return when(result.status) {
            Resource.Status.SUCCESS -> Resource.success(newState)
            else -> Resource.error(result.code, result.message, newState)
        }
    }

    suspend fun resumeCleaning(robot: Robot, params: HashMap<String, String>? = null): Resource<RobotState> {
        val command = JSONObject().apply {
            put("reqId", "77")
            put("cmd", "resumeCleaning")
        }
        val result = nucleoRepository.executeCommand(robot, command)

        val newState = RobotState(serial = robot.serial).apply {
            loadData(result.data)
        }

        return when(result.status) {
            Resource.Status.SUCCESS -> Resource.success(newState)
            else -> Resource.error(result.code, result.message, newState)
        }
    }

    suspend fun returnToBase(robot: Robot, params: HashMap<String, String>? = null): Resource<RobotState> {
        val command = JSONObject().apply {
            put("reqId", "77")
            put("cmd", "sendToBase")
        }

        val result = nucleoRepository.executeCommand(robot, command)

        val newState = RobotState(serial = robot.serial).apply {
            loadData(result.data)
        }

        return when(result.status) {
            Resource.Status.SUCCESS -> Resource.success(newState)
            else -> Resource.error(result.code, result.message, newState)
        }
    }

    open suspend fun startCleaning(robot: Robot, params: CleaningParams? = null): Resource<RobotState> {
        // override default values, with user values
        // discard unsupported values
        if (params != null) {
            if (params.category != null) {
                cleaningType = params.category!!
            }
            if (params.mode != null) {
                cleaningMode = params.mode!!
            }
            if (params.modifier != null) {
                cleaningModifier = params.modifier!!
            }
            if (params.navigationMode != null) {
                cleaningNavigationMode = params.navigationMode!!
            }
            if (params.spotSize != null) {
                cleaningAreaHeight = params.spotSize!!
                cleaningAreaWidth = params.spotSize!!
            }
            if (params.zoneId != null) {
                boundaryId = params.zoneId!!
            }
            if (params.zonesId != null && params.zonesId?.isNotEmpty() == true) {
                boundaryIds = params.zonesId!!
            }
        }

        val command = JSONObject()
        try {
            command.put("reqId", "77")
            command.put("cmd", "startCleaning")

            val commandParams = JSONObject()
            commandParams.put(Nucleo.CLEANING_CATEGORY_KEY, cleaningType.value)
            if (isEcoModeSupported && isTurboModeSupported) {
                commandParams.put(Nucleo.CLEANING_MODE_KEY, cleaningMode.value)
            }
            if (isCleaningFrequencySupported) {
                commandParams.put(Nucleo.CLEANING_MODIFIER_KEY, cleaningModifier.value)
            }

            // this is an exception house basic-1 support only frequency normal fixed
            if (this is HouseCleaningBasic1Service) {
                commandParams.put(Nucleo.CLEANING_MODIFIER_KEY, CleaningModifier.NORMAL.value)
            }

            if (isExtraCareModeSupported) {
                commandParams.put(Nucleo.CLEANING_NAVIGATION_MODE_KEY, cleaningNavigationMode.value)
            }

            if (!boundaryId.isNullOrEmpty()) {
                commandParams.put(Nucleo.ZONE_BOUNDARY_ID_KEY, boundaryId)
            }

            if (boundaryIds.isNotEmpty()) {
                commandParams.put(Nucleo.ZONE_BOUNDARIES_ID_KEY, JSONArray(boundaryIds))
            }

            if (isCleaningAreaSupported) {
                commandParams.put(Nucleo.CLEANING_AREA_SPOT_HEIGHT_KEY, cleaningAreaHeight)
                commandParams.put(Nucleo.CLEANING_AREA_SPOT_WIDTH_KEY, cleaningAreaWidth)
            }
            command.put("params", commandParams)
        } catch (e: JSONException) {
            Log.e(TAG, "Exception", e)
        }

        val result = nucleoRepository.executeCommand(robot, command)

        val newState = RobotState(serial = robot.serial).apply {
            loadData(result.data)
        }

        return when(result.status) {
            Resource.Status.SUCCESS -> Resource.success(newState)
            else -> Resource.error(result.code, result.message, newState)
        }
    }

    companion object {
        private const val TAG = "CleaningService"
    }
}
